test/persist-catalog-migration/mzcompose.py

```diff
@@ -0,0 +1,198 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from textwrap import dedent
+
+from materialize.mzcompose.composition import Composition
+from materialize.mzcompose.services.cockroach import Cockroach
+from materialize.mzcompose.services.materialized import Materialized
+from materialize.mzcompose.services.minio import Minio
+from materialize.mzcompose.services.testdrive import Testdrive
+
+SERVICES = [
+    Minio(setup_materialize=True),
+    Cockroach(setup_materialize=True),
+    Materialized(catalog_store="stash"),
+    Testdrive(no_reset=True),
+]
+
+
+def workflow_default(c: Composition) -> None:
+    for i, name in enumerate(c.workflows):
+        if name == "default":
+            continue
+        with c.test_case(name):
+            c.workflow(name)
+
+
+def workflow_test_migration_and_rollback(c: Composition) -> None:
+    c.down(destroy_volumes=True)
+    c.up("minio", "cockroach")
+    c.up("testdrive", persistent=True)
+
+    # Create some objects.
+    c.up("materialized")
+    create_objects(c, 0)
+
+    # Switch to persist catalog.
+    c.testdrive(
+        input=dedent(
+            """
+    $ postgres-execute connection=postgres://mz_system@materialized:6877/materialize
+    ALTER SYSTEM SET catalog_kind TO 'persist'
+    """
+        )
+    )
+    c.kill("materialized")
+
+    # Reboot and check that objects still exists.
+    c.up("materialized")
+    check_objects(c, 0)
+    # Create some more objects.
+    create_objects(c, 1)
+
+    # Rollback to stash.
+    c.testdrive(
+        input=dedent(
+            """
+    $ postgres-execute connection=postgres://mz_system@materialized:6877/materialize
+    ALTER SYSTEM SET catalog_kind TO 'stash'
+    """
+        )
+    )
+    c.kill("materialized")
+
+    # Reboot and check that objects still exists.
+    c.up("materialized")
+    check_objects(c, 0)
+    check_objects(c, 1)
+
+
+# Tests that Materialize doesn't crash due to the epoch going backwards after a migration.
+def workflow_test_epoch_migration(c: Composition) -> None:
+    reboots = 2
+    c.down(destroy_volumes=True)
+    c.up("minio", "cockroach")
+    c.up("testdrive", persistent=True)
+
+    # Switch to emergency stash catalog so only the stash epoch is incremented.
+    c.up("materialized")
+    c.testdrive(
+        input=dedent(
+            """
+    $ postgres-execute connection=postgres://mz_system@materialized:6877/materialize
+    ALTER SYSTEM SET catalog_kind TO 'emergency-stash'
```

Comment:
Would it be more realistic to set this during startup of materialized already? You could use `c.override(Materialized(...)):` for that.

Reply:
Yeah, just updated to do this.